src/message_metadata.cc

```diff
@@ -0,0 +1,124 @@
+    }
+
+    //---------------------------------------------------------------------
+    //---------------------------------------------------------------------
+	MessageElementMetadata::MessageElementMetadata(IMessageElementMetadataOwner* owner, LVMesageElementMetadata* lvElement, int metadataVersion)
+    {
+        _owner = owner;
+        clusterOffset = 0;
+        embeddedMessageName = GetLVString(lvElement->embeddedMessageName);
+        isRepeated = lvElement->isRepeated;
+        protobufIndex = lvElement->protobufIndex;
+        type = (LVMessageMetadataType)lvElement->valueType;
+
+        if (metadataVersion >= 2)
+        {
+            fieldName = GetLVString(lvElement->fieldName);
+            isInOneof = lvElement->isInOneof;
+            oneofContainerName = GetLVString(lvElement->oneofContainerName);
+        }
+        else
+        {
+            isInOneof = false;
+        }
+
+        // Calculate well known type last after initializing all other fields.
+        auto it = _wellKnownTypeFunctionMap.find(embeddedMessageName);
+        if (it != _wellKnownTypeFunctionMap.end())
+        {
+            wellKnownType = it->second(*this);
+        }
+        else
+        {
+            wellKnownType = wellknown::Types::None;
+        }
+    }
+
+    //---------------------------------------------------------------------
+    //---------------------------------------------------------------------
+    MessageMetadata::MessageMetadata(IMessageElementMetadataOwner* metadataOwner, LVMessageMetadata* lvMetadata)
+    {
+        auto name = GetLVString(lvMetadata->messageName);
+        messageName = name;
+        typeUrl = name;
+        clusterSize = 0;
+        alignmentRequirement = 0;
+
+        if (lvMetadata->elements != nullptr)
+        {
+            // byteAlignment for LVMesageElementMetadata would be the size of its largest element which is a LStrHandle
+            auto lvElement = (LVMesageElementMetadata*)(*lvMetadata->elements)->bytes(0, sizeof(LStrHandle));
+            auto metadataVersion = 1;
+            InitializeElements(metadataOwner, lvElement, (*lvMetadata->elements)->cnt, metadataVersion);
+        }
+    }
+
+    //---------------------------------------------------------------------
+    //---------------------------------------------------------------------
+    MessageMetadata::MessageMetadata(IMessageElementMetadataOwner* metadataOwner, LVMessageMetadata2* lvMetadata)
+    {
+        auto name = GetLVString(lvMetadata->messageName);
+        messageName = name;
+        typeUrl = GetLVString(lvMetadata->typeUrl);
+        clusterSize = 0;
+        alignmentRequirement = 0;
+
+        if (lvMetadata->elements != nullptr)
+        {
+            // byteAlignment for LVMesageElementMetadata would be the size of its largest element which is a LStrHandle
+            auto lvElement = (LVMesageElementMetadata*)(*lvMetadata->elements)->bytes(0, sizeof(LStrHandle));
```

Comment:
Should this use `alignof(LVMesageElementMetadata)`? https://en.cppreference.com/w/cpp/language/alignof

Reply:
Updated.